src/rust/ide/lib/view/src/graph_editor/component/edge.rs

```diff
@@ -884,7 +1665,206 @@ impl EdgeModelData {
+    /// euclidean distance between point and `Input`/`Output`.
+    fn closest_end_for_point(&self, point:Vector2<f32>) -> PortType {
+        let target_position = self.target_position.get().xy();
+        let source_position = self.position().xy() - Vector2(0.0,self.source_height.get() / 2.0);
+        let target_distance = (point - target_position).norm();
+        let source_distance = (point - source_position).norm();
+        if source_distance > target_distance { PortType::OutputPort }
+        else                                 { PortType::InputPort }
+    }
+
+    /// Indicates whether the height difference between input and output is too small to  use the
+    /// y value to assign the `EndDesignation` for a given point.
+    fn input_and_output_y_too_close(&self) -> bool {
+        let target_y = self.position().y;
+        let source_y = self.target_position.get().y;
+        let delta_y  = target_y - source_y;
+        delta_y > 0.0 && delta_y < MIN_SOURCE_TARGET_DIFFERENCE_FOR_Y_VALUE_DISCRIMINATION
+    }
+
+    // FIXE: several lines are too long.
+    /// Return the correct cut angle for the given `shape_id` at the `position` to highlight the
+    /// `target_end`. Will return `None` if the `shape_id` is not a valid sub-shape of this edge.
+    fn cut_angle_for_shape
+    (&self, shape_id:display::object::Id, position:Vector2<f32>, target_end: PortType) -> Option<f32> {
+        let shape      = self.get_shape(shape_id)?;
+        let shape_role = self.get_shape_role(shape_id)?;
+
+        let cut_angle_correction = self.get_cut_angle_correction(shape_role);
+        let target_angle         = self.get_target_angle(target_end);
+
+        let base_rotation = shape.display_object().rotation().z + 2.0 * RIGHT_ANGLE;
+        let shape_normal  = shape.normal(position).angle();
+        Some(shape_normal - base_rotation + cut_angle_correction + target_angle)
+    }
+
+    /// Return the cut angle value needed to highlight the given end of the shape. This takes into
+    /// account the current layout.
+    fn get_target_angle(&self, target_end: PortType) -> f32 {
+        let output_on_top = self.layout_state.get().is_output_above_input();
+        match (output_on_top,target_end) {
+            (false, PortType::InputPort)  => 2.0 * RIGHT_ANGLE,
+            (false, PortType::OutputPort) => 0.0,
+            (true, PortType::InputPort)   => 0.0,
+            (true, PortType::OutputPort)  => 2.0 * RIGHT_ANGLE,
+        }
+    }
+
+    /// These corrections are needed as sometimes shapes are in places that lead to inconsistent
+    /// results, e.g., the side line leaving the node from left/right or right/left. The shape
+    /// itself does not have enough information about its own placement to determine which end
+    /// is pointed towards the `Target` or `Source` part of the whole edge. So we need to account
+    /// for these here based on the specific layout state we are in.
```

Comment:
Could you please describe an example case preferably with ascii-art here? It would be sooooo helpful to understand it. Now I understand what is happening here, but I bet in couple of months this will be hard to follow.